### PR TITLE
libnss: add nss-tools 3.43 a prereq for libreswan

### PR DESCRIPTION
**Pull request by lucize**
Signed-off-by: Lucian Cristian <lucian.cristian@gmail.com>

Maintainer: me 
LEDE r5217-098afa1e1b
Compile tested: mips_24kc (glibc/musl)
Compile/Run tested: x86_64, x86, arm_cortex-a8_vfpv3 (glibc/musl) 

Description:
A dependency for the upcoming libreswan package, a bit delayed by some issues on shutdown if using musl
depends on #8561

### CODE REVIEW

**Comment by champtar** [comment]
I though openssl was big, but libnss is even bigger
1120435	bin/packages/mips_24kc/packages/libnss_3.33-1_mips_24kc.ipk
+ dependencies:
339316	bin/packages/mips_24kc/packages/libsqlite3_3200100-1_mips_24kc.ipk
375368	bin/targets/ar71xx/generic/packages/libstdcpp_5.5.0-1_mips_24kc.ipk
...

Maybe you can do more granular packaging, as I don't think libreswan need libsmime3.so for example

**Comment by lucize** [comment]
[this ](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/An_overview_of_NSS_Internals) article contains  `you may initialize NSS in a no-database mode`
I'll try to find out more about this

File: libs/nss/Makefile [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+		TARGET=$(TARGET_CC) TARGETCXX=$(TARGET_CXX) CPU_ARCH=$(ARCH) MUSL=$(if $(CONFIG_LIBC_USE_GLIBC),0,1))
+endef
+
+define Package/libnss
+  SECTION:=libs
+  SUBMENU:=SSL
+  CATEGORY:=Libraries
+  TITLE:=Library Network Security Services (NSS)
+  URL:=http://www.mozilla.org/projects/security/pki/nss/
+  DEPENDS:=+libpthread +librt +zlib +libstdcpp +libsqlite3
```
Comment (neheb) [inline libs/nss/Makefile]: is uclibc++ possible here?

Comment (lucize) [reply libs/nss/Makefile]: I think is not needed anymore but I got a linking problem with libreswan that I didn't saw last time, need more testing

File: libs/nss/Makefile [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=nss
+PKG_VERSION:=3.42.1
+PKG_RELEASE:=1
+PKG_MAINTAINER:=Lucian Cristian <lucian.cristian@gmail.com>
+PKG_LICENCE:=GPL-2.0+
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION)-with-nspr-4.20.tar.gz
+PKG_SOURCE_URL:=https://ftp.mozilla.org/pub/security/nss/releases/NSS_3_42_1_RTM/src/
```
Comment (diizzyy) [inline libs/nss/Makefile]: Please use 
https://download.cdn.mozilla.net/pub/security/nss/releases/NSS_3_42_1_RTM/src/ and 
https://archive.mozilla.org/pub/security/nss/releases/NSS_3_42_1_RTM/src/ instead

File: libs/nss/Makefile [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+endif
+
+define Build/Compile
+	(cd $(PKG_BUILD_DIR)/nss; \
+		NSPR_CONFIGURE_OPTS="--build=$(GNU_HOST_NAME) \
+				     --host=$(GNU_HOST_NAME) \
+				     --target=$(REAL_GNU_TARGET_NAME) \
+				     --disable-debug \
+				     --enable-optimize \
+				     --enable-strip \
```
Comment (diizzyy) [inline libs/nss/Makefile]: Is there a need to override the toolchain?

File: libs/nss/patches/010-create_openwrt_target.patch [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
++include $(CORE_DEPTH)/coreconf/UNIX.mk
++
++RANLIB = ranlib
++CC = $(TARGET)
++DEFAULT_COMPILER= $(TARGET)
++CCC = $(TARGETCXX)
++CROSS_COMPILE = 1
++NATIVE_CC = gcc
++DEFINES += -DHAVE_POSIX_FALLOCATE=0  -D_REENTRANT
++OPTIMIZER = -Os -gdwarf-2
```
Comment (diizzyy) [inline libs/nss/patches/010-create_openwrt_target.patch]: If possible strip all additional flags and add use the TARGET_CFLAGS variable instead

**Comment by diizzyy** [comment]
Ca we also break out the certificates?
https://www.freshports.org/security/ca_root_nss/

File: libs/nss/Makefile (hunk 1 above)
Comment (neheb) [reply libs/nss/Makefile]: Ah the catch is that everything must include uclibc++.mk. you can't link between uclibc++ and libstdcpp.

Comment (neheb) [reply libs/nss/Makefile]: For example this: https://github.com/neheb/packages/commit/33b1429019b0182c84d3ec3d8a1e10a4c65bc09e

Only ibrcommon depends on any kind of libc++ (technically they all do) but uclibc++.mk must still be included in all of them.

**Comment by lucize** [comment]
@diizzyy 
this archive has nss and nspr all in one, but I started the other way to build them separately, nspr is working but nss (for now) `fails with x86_64-openwrt-linux-musl/bin/ar: invalid option -- 'L'` and didn't find a quick answer, also I'm trying not to add an openwrt.mk target at all

File: libs/nss/Makefile [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+# Copyright (C) 2019 Lucian Cristian
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=nss
+PKG_VERSION:=3.43
+_PKG_VERSION:=3_43
```
Comment (diizzyy) [inline libs/nss/Makefile]: This is ugly, remove and hardset it instead

File: libs/nss/Makefile [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+PKG_RELEASE:=1
+PKG_MAINTAINER:=Lucian Cristian <lucian.cristian@gmail.com>
+PKG_LICENCE:=GPL-2.0+
```
Comment (diizzyy) [inline libs/nss/Makefile]: Not GPL-2.0+
https://hg.mozilla.org/projects/nss/file/NSS_3_43_BRANCH/COPYING

File: libs/nss/Makefile (hunk 5 above)
Comment (lucize) [reply libs/nss/Makefile]: I think I saw somewhere a replace function so I only have to change in one place, I'll look into this

File: libs/nss/Makefile [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://archive.mozilla.org/pub/security/$(PKG_NAME)/releases/NSS_$(_PKG_VERSION)_RTM/src
```
Comment (diizzyy) [inline libs/nss/Makefile]: Add https://download.cdn.mozilla.net/ for primary URL

File: libs/nss/Makefile [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+export NSS_ENABLE_WERROR=0
+export NSS_DISABLE_GTESTS=1
+export NSS_USE_SYSTEM_SQLITE=1
+export OS_TARGET=Linux
+export OS_ARCH=Linux
+export OS_TEST=$(ARCH)
+export CPU_ARCH=$(ARCH)
+export NSPR_INCLUDE_DIR=$(STAGING_DIR)/usr/include/nspr
+export SEED_ONLY_DEV_URANDOM=1
+#size optimisation
```
Comment (diizzyy) [inline libs/nss/Makefile]: This should be a configuration option and you probably want to disable MIPS16 due to performance issues

File: libs/nss/Makefile [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+  Network Security Services (NSS) is a set of libraries designed to support 
+  cross-platform development of security-enabled client and server applications.
+  Applications built with NSS can support SSL v2 and v3, TLS, PKCS 5, PKCS 7,
+  PKCS 11, PKCS 12, S/MIME, X.509 v3 certificates, and other security standards.
+endef
+
+CONFIGURE_PATH = ./nss
+MAKE_PATH = ./nss
+
+ifeq ($(ARCH),aarch64)
```
Comment (neheb) [inline libs/nss/Makefile]: are these really needed?

Comment (lucize) [reply libs/nss/Makefile]: if there is another way to pass `USE_64=1` for 64 bit targets then I'll use that, otherwise compile will fail

Comment (neheb) [reply libs/nss/Makefile]: Maybe use some sh magic to match against 64 in $(ARCH).

Comment (lucize) [reply libs/nss/Makefile]: Yes, I think is safe to assume that if arch contains 64 is a 64 bit target

**Comment by lucize** [comment]
runtested on x86 x86_64, only tested PSK ipsec,
no speed improvements where observed between Os and O2, anyway adding the TARGET_CFLAGS overrides O2 option

**Comment by neheb** [comment]
Not building. Maybe needs a rebase?

**Comment by diizzyy** [comment]
Looks like there's some kind of CFLAGS pollution too
```mips-openwrt-linux-musl-gcc -o build_dir/quickder.o -c -Os -fpic -Os```

**Comment by lucize** [comment]
> -Os -fpic -Os
@diizzyy  first one is from exported build opt and the second one is from target_cflags
so what do you propose, to remove the build opt setting and rely only on target_cflags set on openwrt ?
@neheb in needs nspr. that was just merged for the build to work

**Comment by lucize** [comment]
reopening doesn't seem to restart the CI build

**Comment by lucize** [comment]
removed the hardcoded O flag, but I don't know how much time we need to wait for the build bots to take the new master into acoount

**Comment by diizzyy** [comment]
You need to rebase against master

**Comment by neheb** [comment]
@lucize click CircleCI details. There's an option to rerun on the left.

**Comment by lucize** [comment]
it's good now

File: libs/nss/Makefile [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+PKG_BUILD_PARALLEL:=0
+
+include $(INCLUDE_DIR)/package.mk
+
+define Package/libnss
+  SECTION:=libs
+  SUBMENU:=SSL
+  CATEGORY:=Libraries
+  TITLE:=Mozilla's SSL and TLS implementation
+  URL:=http://www.mozilla.org/projects/security/pki/nss/
```
Comment (neheb) [inline libs/nss/Makefile]: HTTPS. URL also needs updating.

File: libs/nss/Makefile [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+  DEPENDS:=+libpthread +librt +libsqlite3 +nspr
+endef
+
+define Package/nss-utils
+  SECTION:=utils
+  CATEGORY:=Utilities
+  TITLE:=Mozilla's SSL and TLS implementation utilities
+  URL:=http://www.mozilla.org/projects/security/pki/nss/
```
Comment (neheb) [inline libs/nss/Makefile]: Same as above

**Comment by lucize** [comment]
@neheb do you think should I try to see if uclibc++ can be used ? I need to try uclibc++ on every dependent package ?

**Comment by neheb** [comment]
Here's the rule: All dependent packages must also be supported by uclibc++. You cannot link between uclibc++ and libstdcpp.

uclibc++ does not support most C++11 features.

**Comment by lucize** [comment]
@neheb it's ready

**Comment by neheb** [comment]
According to https://wiki.mozilla.org/NSS/C++ , uclibc++ cannot be used. nullptr is not available in uclibc++.

**Pull request merged**